prometheus/push.go

```diff
@@ -19,47 +19,153 @@
 
 package prometheus
 
-// Push triggers a metric collection by the default registry and pushes all
-// collected metrics to the Pushgateway specified by url. See the Pushgateway
-// documentation for detailed implications of the job and instance
-// parameter. instance can be left empty. You can use just host:port or ip:port
-// as url, in which case 'http://' is added automatically. You can also include
-// the schema in the URL. However, do not include the '/metrics/jobs/...' part.
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+
+	"github.com/prometheus/common/expfmt"
+	"github.com/prometheus/common/model"
+)
+
+// Push works like PushFrom but collects the pushed metrics from the
+// DefaultRegistry.
+func Push(job string, grouping map[string]string, url string) error {
+	return PushFrom(DefaultRegistry, job, grouping, url)
+}
+
+// PushAdd works like PushAddFrom but collects the pushed metrics from the
+// DefaultRegistry.
+func PushAdd(job string, grouping map[string]string, url string) error {
+	return PushAddFrom(DefaultRegistry, job, grouping, url)
+}
+
+// PushFrom triggers a metric collection by the provided registry and pushes all
+// collected metrics to the Pushgateway specified by url, using the provided job
+// name and the (optional) further grouping labels (the grouping map may be
+// nil). See the Pushgateway documentation for detailed implications of the job
+// and other grouping labels. Neither the job name nor any grouping label value
+// may contain a "/". The metrics pushed must not contain a job label of their
+// own nor any of the grouping labels.
+//
+// You can use just host:port or ip:port as url, in which case 'http://' is
+//added automatically. You can also include the schema in the URL. However, do
+//not include the '/metrics/jobs/...' part.
 //
-// Note that all previously pushed metrics with the same job and instance will
-// be replaced with the metrics pushed by this call. (It uses HTTP method 'PUT'
-// to push to the Pushgateway.)
-func Push(job, instance, url string) error {
-	return defRegistry.Push(job, instance, url, "PUT")
+// Note that all previously pushed metrics with the same job and other grouping
+// labels will be replaced with the metrics pushed by this call. (It uses HTTP
+// method 'PUT' to push to the Pushgateway.)
+func PushFrom(r Registry, job string, grouping map[string]string, url string) error {
```

Comment:
> It helps if we can encapsulate and organise related behaviour. As it stands all these top-level functions are going to make the API very unwieldy.

You said that before. I have understood you. As said before, I completely agree that many top-level identifiers are unwieldy. There is no point in repeating that again and again. However, as I have also said before, I also think that a `struct` in Go is not meant to organize related behavior, and doing so doesn't solve the problem. Just repeating what you said before won't change my opinion here. It merely triggers me to repeat what I said, too.

To add something of new value, hopefully:

In most cases, users of a PGW push exactly once (at the end of their batch job). With the current approach, that's one call. With a struct wrapping, that becomes a two step process: Create an instance of the struct, then call a method. In the code, we would replace four top-level identifiers by one top-level identifier (the struct) with four methods, which results in five entries in the Godoc. Even before, all four functions were next to each other in the Godoc and quite obviously related by naming. I'd argue the situation is now worse. One more thing to mentally parse, and one more step in pushing.

Other suggestion: Since I have removed the global push function for the default registry, the push stuff is, for a change, a part we can move into its own package right now. How about doing that? There is a bit of friction, because it would be the first package, but it's a first step, and then pushing will stay a one-step procedure, and it will be really idiomatic:

```Go
import ".../prometheus/push"

push.Registry(reg, "myjob", nil, url)
push.Collectors("myjob", nil, url, c1, c2)
```

Perhaps move the registry argument to the end to mirror the collector location? (The collectors have to be at the end to make them variadic.)